File: pure_cnn_react/src/model/maxpool_layer.js
import Layer from './layer';
import Mat from '../util/mat';
import Shape from '../util/mat_shape';
import {INIT_CHARS, INIT_ZEROS, LAYER_MAXPOOL} from '../util/constant';


export default class MaxPoolLayer extends Layer {
    constructor(name, pool_width, pool_height, pool_stride_x, pool_stride_y) {
        super(name, 0);
        this.type = LAYER_MAXPOOL;
        this.pool_width = pool_width;
        this.pool_height = pool_height;
        this.pool_stride_x = pool_stride_x;
        this.pool_stride_y = pool_stride_y;
        this.max_info = undefined
    }

    set_input_layer(inputLayer) {
        super.set_input_layer(inputLayer);

        this.output_width = Math.floor((this.input_shape.width - this.pool_width) / this.pool_stride_x + 1);
        this.output_height = Math.floor((this.input_shape.height - this.pool_height) / this.pool_stride_y + 1);
        this.output_depth = this.input_shape.depth;

        this.output_shape = new Shape(this.output_width, this.output_height, this.output_depth);
    }

    forward() {
        this.input = this.pre_layer.output;
        this.output = new Array(this.network.mini_batch_size);
        this.max_info = new Array(this.network.mini_batch_size);

        for (let i = 0; i < this.network.mini_batch_size; ++i) {
            this.max_info[i] = new Mat(this.output_shape, INIT_CHARS);
            this.output[i] = new Mat(this.output_shape, INIT_ZEROS);

            for (let out_d = 0; out_d < this.output_depth; out_d++) {
                for (let out_y = 0; out_y < this.output_height; out_y++) {
                    for (let out_x = 0; out_x < this.output_width; out_x++) {
                        let max_value = 0;
                        let max_index = [0, 0, 0];
                        for (let input_y = 0; input_y < this.pool_height; input_y++) {
                            let stride_y = out_y * this.pool_stride_y + input_y;
                            for (let input_x = 0; input_x < this.pool_width; input_x++) {
                                let stride_x = out_x * this.pool_stride_x + input_x;
                                let compare_input = this.input[i].get_value_by_coordinate(stride_x, stride_y, out_d);
                                if (compare_input >= max_value) {
                                    max_value = compare_input;
                                    max_index = [stride_x, stride_y, out_d];
                                }
                            }
                        }
                        this.max_info[i].set_value_by_coordinate(out_x, out_y, out_d, max_index.join("-"));
                        this.output[i].set_value_by_coordinate(out_x, out_y, out_d, max_value);
                    }
                }
            }
        }
    }

    backward() {
        let next_layer = this.next_layer;
        this.back_error = new Array(this.network.mini_batch_size);

        for (let i = 0; i < this.network.mini_batch_size; ++i) {
            this.back_error[i] = new Mat(this.input_shape, INIT_ZEROS);

            let max_infos = this.max_info[i].get_value();
            for (let mi = 0; mi < max_infos.length; mi++) {
                let info = max_infos[mi].split("-");
                let input_x = parseInt(info[0], 10);
                let input_y = parseInt(info[1], 10);
                let input_d = parseInt(info[2], 10);
                this.back_error[i].set_value_by_coordinate(
                    input_x,
                    input_y,
                    input_d,
                    next_layer.back_error[i].get_value_by_coordinate(
                        Math.floor(input_x / this.pool_width),
                        Math.floor(input_y / this.pool_height),
                        input_d));
            }
        }
    }
}

File: pure_cnn_py/model/maxpool_layer.py
from pure_cnn_py.model.layer import Layer
from pure_cnn_py.util.mat import Mat
from pure_cnn_py.util.mat_shape import MatShape
from pure_cnn_py.util.constant import INIT_ZEROS, LAYER_MAXPOOL


class MaxpoolLayer(Layer):
    def __init__(self, name, pool_width, pool_height, pool_stride_x, pool_stride_y, network):
        super(MaxpoolLayer, self).__init__(name, 0)
        self.layer_type = LAYER_MAXPOOL

        self.pool_width = pool_width
        self.pool_height = pool_height
        self.pool_stride_x = pool_stride_x
        self.pool_stride_y = pool_stride_y

        self.network = network
        self.input = None
        self.output = None

        self.output_width = None
        self.output_height = None
        self.output_depth = None
        self.output_shape = None

        self.max_info = None

        self.back_error = None

        self.pool_max_activation_index = None

    def set_input_layer(self, input_layer):
        super(MaxpoolLayer, self).set_input_layer(input_layer)
        self.output_width = (self.input_shape.width - self.pool_width) // self.pool_stride_x + 1
        self.output_height = (self.input_shape.height - self.pool_height) // self.pool_stride_y + 1
        self.output_depth = self.input_shape.depth

        self.output_shape = MatShape(self.output_width, self.output_height, self.output_depth)

    def forward(self):
        self.input = self.pre_layer.output
        self.output = [0] * self.network.mini_batch_size
        self.max_info = [0] * self.network.mini_batch_size

        for i in range(0, self.network.mini_batch_size):
            self.output[i] = Mat(self.output_shape, INIT_ZEROS)
            self.max_info[i] = Mat(self.output_shape, INIT_ZEROS)

            for out_d in range(0, self.output_depth):
                for out_y in range(0, self.output_height):
                    for out_x in range(0, self.output_width):
                        max_value = 0
                        max_index = [0, 0, 0]
                        for input_y in range(0, self.pool_height):
                            stride_y = out_y * self.pool_stride_y + input_y
                            for input_x in range(0, self.pool_width):
                                stride_x = out_x * self.pool_stride_x + input_x
                                compare_input = self.input[i].get_value_by_coordinate(stride_x, stride_y, out_d)
                                if compare_input >= max_value:
                                    max_value = compare_input
                                    max_index = [stride_x, stride_y, out_d]

                        self.max_info[i].set_value_by_coordinate(out_x, out_y, out_d,
                                                                 "-".join([str(i) for i in max_index]))
                        self.output[i].set_value_by_coordinate(out_x, out_y, out_d, max_value)

    def backward(self):
        next_layer = self.next_layer
        self.back_error = [0] * self.network.mini_batch_size

        for i in range(0, self.network.mini_batch_size):
            self.back_error[i] = Mat(self.input_shape, INIT_ZEROS)
            for mi in self.max_info[i].get_value():
                input_x, input_y, input_d = [int(a) for a in mi.split("-")]
                self.back_error[i].set_value_by_coordinate(input_x,
                                                           input_y,
                                                           input_d,
                                                           next_layer.back_error[i].get_value_by_coordinate(
                                                               input_x // self.pool_width,
                                                               input_y // self.pool_height,
                                                               input_d))
